### ISSUE TITLE
Consider simplifying SpanAttribute types - use Swift types?

### ISSUE THREAD

**Issue by pokryfka** [slashmo/gsoc-swift-tracing#83]
Per [OT](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes):

> The Span interface MUST provide:
> An API to set a single Attribute where the attribute properties are passed as arguments. This MAY be called SetAttribute.
> To avoid extra allocations some implementations may offer a separate API for each of the possible value types.

`TracingInstrument` defines two public type: `SpanAttribute` and `SpanAttributes` which provide A LOT more and need to be implemented by any `TracingInstrument` implementation either by:

- embracing them and adding dependency on the types even if they not directly match their own internal definition of "an attribute" (see XRays annotation and metadata); or
- adding lots of bridging and copying data back and forward in the process

to be honest I think `SpanAttribute` and `SpanAttributes` are a nice example of Swift type safety but its a bit of an hassle to adapt to

in my original XRay Segment implementation I defined only setters (no getters) for XRay Annotation and decided to expose only Swift types:

```swift
    internal typealias Annotations = [String: AnnotationValue]

    internal enum AnnotationValue {
        case string(String)
        case integer(Int)
        case float(Float)
        case bool(Bool)
    }

    private func setAnnotations(_ newElements: Annotations) {
        lock.withWriterLockVoid {
            for (k, v) in newElements {
                _annotations.updateValue(v, forKey: k)
            }
        }
    }

    public func setAnnotation(_ value: String, forKey key: String) {
        setAnnotations([key: .string(value)])
    }
    public func setAnnotation(_ value: Bool, forKey key: String) {
        setAnnotations([key: .bool(value)])
    }

    public func setAnnotation(_ value: Int, forKey key: String) {
        setAnnotations([key: .integer(value)])
    }

    public func setAnnotation(_ value: Float, forKey key: String) {
        setAnnotations([key: .float(value)])
    }
```

added a getter and a function to remove a value trying to conform to `Span` type but that's still not enough,
I need to expose a (copy) of the collection with all the annotations and a function with a callback modifying all the annotations.

Which OT requirement does it fulfils and what is a use case for that?

I think in a typical use case user creates a span, adds a few attributes (perhaps overriding previous ones) and that's it really.

**Comment by slashmo** [comment]
Thanks for this 👍

## SpanAttribute

That's a good point. What if we'd additionally provide methods for accepting the different supported Swift types in a protocol extension?

## SpanAttributes

In Otel, there's an additional "SHOULD" to preserve the order of attributes: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes

> Attributes SHOULD preserve the order in which they're set.

That's why we decided against simply using a dictionary as the public interface for `SpanAttributes`. Even though the wrapper currently still uses a dictionary for internal storage, that way we still have the option to change it later without breaking API.

**Comment by ktoso** [comment]
Optimal would be to allow impls to conform to that "should" clause but not force excessive copying to others... without libs knowing about the tracing types that may be hard though 🤔 Open to ideas

**Comment by pokryfka** [comment]
> Attributes SHOULD preserve the order in which they're set.

this is a valid requirement (though to be honest it will be difficult to enforce it outside of a specific implementation of an `TracingInstrument`.

my complains are: which requirement and/or use case requires every `TracingInstrument` to:

- return a copy of all the attributes; in fact I think there is NO requirement (as defined by OT) to "get" an attribute value at all, only to "set" them 
- provide a callback to modify their internal values?
- even if it is required to return a collection of all the attributes (to be honest I am not sure which use cases does it address), what's wrong with Swift Collection types?

`SpanAttributes` and `SpanAttribute` provide a nice and type safe implementation ON TOP of the requirements of OT SpanAttribute; nice as long as it matches directly the needs of an `TracingInstrument` otherwise it still forces it to implement because the types are part of the protocol

TL;DR

rather than defining (minimal) requirements of a `TracingInstrument` and let them implement it: 

> An API to set a single Attribute where the attribute properties are passed as arguments. 
> This MAY be called SetAttribute. To avoid extra allocations some implementations may offer a separate API for each of the possible value types.

> An Attribute is defined by the following properties:
> 
> (Required) The attribute key, which MUST be a non-null and non-empty string.
> (Required) The attribute value, which is either:
> A primitive type: string, boolean or numeric.
> An array of primitive type values. The array MUST be homogeneous, i.e. it MUST NOT contain values of different types.

`SpanAttributes` and `SpanAttribute` provide AN opinionated solution of how to implement them and force all `TracingInstrument` to use it which may or may not be convinient

**Comment by pokryfka** [comment]
proposition for further discussion based on [OT Span](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span)

including:

> After the Span is created, it SHOULD be possible to change the its name, set its Attributes, and add Links and Events. 

IMHO changing the name is arguable but but its not an issue in general

> Spans are not meant to be used to propagate information within a process. To prevent misuse, implementations SHOULD NOT provide access to a Span's attributes besides its SpanContext.


```swift
public protocol Span {

   /// An API that returns the SpanContext for the given Span.
   /// The returned value may be used even after the Span is finished.
   /// The returned value MUST be the same for the entire Span lifetime.
   /// This MAY be called GetContext
    var context: Baggage.BaggageContext { get }

    var isRecording: Bool { get }

    // An API to set a single Attribute where the attribute properties are passed as arguments.
    // This MAY be called SetAttribute.
    // To avoid extra allocations some implementations may offer a separate API for each of the possible value types.

    mutating func setAttribute(_ value: String, forKey key: String)
    mutating func setAttribute(_ value: Bool, forKey key: String)
    mutating func setAttribute(_ value: [String], forKey key: String)
    // ...
    
    mutating func addEvent(_ event: Instrumentation.SpanEvent)

    mutating func setStatus(_ status: Instrumentation.SpanStatus)

    mutating func addLink(_ link: Instrumentation.SpanLink)

    mutating func updateName(_ name: String)

    mutating func end()
    
    /// To facilitate recording an exception languages SHOULD provide a RecordException convenience method.
    /// The signature of the method is to be determined by each language and can be overloaded as appropriate.
    /// The method MUST record an exception as an Event with the conventions outlined in the exception semantic conventions document.
    mutating func recordException(_ error: Swift.Error)
    
    }
```

note I did not expose `Timestamp` type which according to OT is optional.

> (Optional) Timestamp to explicitly set the end timestamp

in the XRay subsegment I made it only internal ans use in unit tests but did not see a reason to let user mess up with time:

```swift
    /// Updates `endTime` of the Segment.
    public func end() {
        try? end(Timestamp())
    }

    internal func end(_ timestamp: Timestamp) throws {
        try lock.withWriterLockVoid {
            switch _state {
            case .inProgress(let startTime):
                guard startTime < timestamp else {
                    throw SegmentError.startedInFuture
                }
                _state = .ended(started: startTime, ended: timestamp)
            case .ended:
                throw SegmentError.alreadyEnded
            case .emitted:
                throw SegmentError.alreadyEmitted
            }
        }
    }
```

**Comment by pokryfka** [comment]
@ktoso @slashmo 
for reference, my comparison of XRay segment and OT Span
https://github.com/slashmo/gsoc-swift-tracing/issues/72#issuecomment-660824378

**Comment by slashmo** [comment]
Closing this as everything agreed upon has been moved into separate issues (see references above).
